scheduler/core/components/ranker/default.py

```diff
@@ -47,6 +53,11 @@ class RankerParameters:
     program_priority: float = 10.0
     priority_factor: float = 8.0
     preimaging_factor: float = 1.25
+    altitude_limits: Dict[str, Dict] = field(default_factory=lambda: _def_alt_limits_site)
```

Comment:
Do we want users of the API to be able to change `altitude_limits`?

If not, then we do this. Note the change in the parameters, which matches the definition above that you use as the default value. (The type hints in dataclasses are really important in Python.)

```python
altitude_limits: Dict[Site, Dict[MinMax, Angle]]  = field(default_factory=lambda: _def_alt_limits_size)
```

If we don't want them to be able to initialize this, we do the same thing, but we change the call to `field` to be:

```python
# Everything else stays the same.
... = field(init=False, ...)
```

Then instead of doing lookups with strings `'min'` and `'max'`, which is inadvisable (you could run into case issues if you forget, and it's just a bad idea to have identical strings littered through the code in general), you'll look up with `MinMax.MIN` and `MinMax.MAX`.

Reply:
Oh, I should give you an example of why we need `field` with `default` (in the case of primitives) and `default_factory` (in the case of more complex types).

Both ChatGPT and I are having trouble finding a use case where `field(..., default=...)` is really necessary outside of exceptional circumstances, so I wouldn't worry about it unless you want to apply other `field` keywords to the field, e.g:

```python
my_field: int = field(init=False, repr=False, default=5)
```

`default_factory` is much more important. To try to give you an idea why, it's because Python handles default complex data structures in a way that seems very unpredictable to most people. Here's some example code:

This is kind of a contrived example, but say you have a simple function to take an item and a list and you want to add the item to the list, and you pass in an empty list as the default:
```python
def add_to_list(item, my_list=[]):
    my_list.append(item)
    return my_list
```

Then, if you do:
```python
my_list = add_to_list('apple') # Expected: ['apple']
my_list2 = add_to_list('banana', my_list) # Expected: ['apple', 'banana']

# In this next case, you'd expect a list just containing cantaloupe due to the empty list default parameter:
my_list3 = add_to_list('cantaloupe')

# Instead, you get that my_list3 is actually ['apple', 'banana', 'cantaloupe']
```

This hopefully gives you a hint why you need to create new objects instead of using defaults, which is what `default_factory` does.

So in the case of your code above, to avoid risking changing the `_def_alt_limits_site`, you use the notation:

```python
... = field(default_factory=lambda: _def_alt_limits_site)
```
instead. Python calls the lambda expression and then makes a clean copy of `_def_alt_limits_site` to use for the dataclass.

This could be even worse if you wanted to, say, build up student grades:

```python
@dataclass
class Grades:
    grades: List[float] = []

    def add_grades(self, grade_batch: List[float]):
        self.grades.extend(grade_batch)
```

Python won't even let you assign `[]` to grades, and gives:

```python
ValueError: mutable default <class 'list'> for field grades is not allowed: use default_factory
```

Reply:
Thanks for the suggestions. I committed new code that implements these. The parameters for the sites can now be given separately and I added error checking. This also requires 
https://github.com/gemini-hlsw/lucupy/pull/100